src/api/orgAuthPlugin.ts

```diff
@@ -0,0 +1,116 @@
+import { getModelForClass } from '@typegoose/typegoose';
+import envVar from 'env-var';
+import type { FastifyInstance, FastifyReply, FastifyRequest } from 'fastify';
+import fastifyPlugin, { type PluginMetadata } from 'fastify-plugin';
+import type { Connection } from 'mongoose';
+
+import { MemberModelSchema, Role } from '../models/Member.model.js';
+import { HTTP_STATUS_CODES } from '../utilities/http.js';
+import type { Result } from '../utilities/result.js';
+import type { PluginDone } from '../utilities/fastify/PluginDone.js';
+
+interface OrgRequestParams {
+  readonly orgName: string;
+  readonly memberId: string;
+}
+
+interface AuthenticatedFastifyRequest extends FastifyRequest {
+  user: { sub: string };
+}
+
+interface AuthorisedFastifyRequest extends AuthenticatedFastifyRequest {
+  isUserAdmin: boolean;
+}
+
+interface AuthorisationGrant {
+  readonly isAdmin: boolean;
+  readonly reason: string;
+}
+
+async function decideAuthorisation(
+  userEmail: string,
+  request: FastifyRequest,
+  dbConnection: Connection,
+  superAdmin?: string,
+): Promise<Result<AuthorisationGrant, string>> {
+  if (superAdmin === userEmail) {
+    return { didSucceed: true, result: { reason: 'User is super admin', isAdmin: true } };
+  }
+
+  const { orgName, memberId } = request.params as Partial<OrgRequestParams>;
+
+  if (orgName === undefined) {
+    return { didSucceed: false, reason: 'Non-super admin tries to access bulk org endpoint' };
+  }
+
+  const memberModel = getModelForClass(MemberModelSchema, {
+    existingConnection: dbConnection,
+  });
+  const member = await memberModel.findOne({ orgName, memberId }).select(['role', 'email']);
+  if (member === null) {
+    return { didSucceed: false, reason: 'User is not a member of the org' };
+  }
+  if (member.role === Role.ORG_ADMIN) {
+    return { didSucceed: true, result: { reason: 'User is org admin', isAdmin: true } };
+  }
+
+  if (memberId === undefined) {
+    return { didSucceed: false, reason: 'User is not accessing a membership' };
+  }
```

Comment:
I think we need to check this before we even try to select the members form the database

Reply:
I considered that but ruled it out: Although it certainly seems like the more intuitive option, it would complicate the code because we'd need to introduce an extra `if`/`else` to make sure org admins are still given access to anything under `/orgs/:orgName` even if there's no `:memberId`.